### ISSUE TITLE
Crowdin

### ISSUE THREAD

**Issue by maybe-eve** [ayust/alpha-guide#8]
Hello, very impressive guide.
Please let me tell you that I'm very glad to know you and your work.

I'm Japanese and I personally have started translation to Japanese, using Crowdin, an online translation platform.
However, while it is very convenient as it has various functions, I find it hard to pull new changes and apply them into Crowdin project.
So I would be appreciate if you could receive its ownership, and integrate it to your GitHub repository so that any commits will be applied automatically.

↓ Please take a look how it works very well! ↓
![Image](http://i.imgur.com/WacOAxr.png)

You can sign up to Crowdin by your GitHub account, which will make it get started quickly.
And once you received the ownership, you can submit [Open Source Project Setup Request](https://crowdin.com/page/open-source-project-setup-request) to use Crowdin for free.
I have already set up the glossary for German, French, Japanese, Russian, and Chinese, including various terms from EVE Online Static Data Export (SDE). It would be very easy to translate EVE-related words.

Once again, great job!
Many Thanks.

**Comment by maybe-eve** [comment]
EDIT: It seems that there is no easy way to transfer project ownership to others.
So if you could create your own translation project within Crowdin, Transifex, or other services, I will send you the glossary data (28407 words!) and my translations.
Thanks!

**Comment by ayust** [comment]
Hi there! I'm definitely willing to look at options for how to translate the guide most effectively. The main trick here is that for me to publish the resulting versions, it's easiest if each language is just a separate Git branch.

I'm not sure if any of these localization platforms support exporting a Git branch with all of the translations in-place, and I'd really prefer not to have to write glue to pull translations back out of a tool if I can avoid it.

In the short term, if you wanted to create a `japanese` branch, I'd be happy to host it at jp.thealphasguide.com or similar.

**Comment by ayust** [comment]
Also, if you want to experiment with how the integration could work - you can always fork this repository to create your own clone of it, and then set up crowdin to integrate with your fork. If you find a way to make it work such that the result is the separate-branches-with-in-place-translation that I described above, let me know!

**Comment by maybe-eve** [comment]
Thanks! I'm now searching for several similar services to compare.
However, I'm now facing the License problem.. I'll submit a new issue about that, so please take a look at that.

**Comment by ayust** [comment]
I've written a guide for now on how I'm currently able to accept translations:
https://github.com/ayust/alpha-guide/wiki/Translating-the-Guide

I'm willing to look at ways to incorporate other tools into this process, but the one written there is the easiest for me.

**Comment by maybe-eve** [comment]
Great! At first I'll submit my translation by that way, then try and consider using tools. 

I'll open a new issue when it is all ready for use and all you have to do is just few clicks, so I close this. I appreciate your continued support!

**Comment by maybe-eve** [comment]
Hi!

Crowdin project is now working!
https://crowdin.com/project/the-alphas-guide-to-the-galaxy
If other translators want to use this as a workbench, please follow this link: https://crowdin.com/project/the-alphas-guide-to-the-galaxy/invite
I've currently set up translation pages (and also EVE glossaries) for Chinese, French, German and Russian.
All traslated files will automatically appear here: https://github.com/maybe-eve/alpha-guide/tree/l10n_master

**Comment by ayust** [comment]
By the way - let me know if you'd like to release a partially-translated version (I've been following your progress, looks like you're about 25-30% done so far?). I'd be happy to put things up at any point and update them from there.

**Comment by maybe-eve** [comment]
Yes, please! I'll finish translating the new-goals directory in few days so it will be good to pre-release.

**Comment by ayust** [comment]
Okay, sounds good.

**Comment by maybe-eve** [comment]
Hello! Here is the partially translated branch.

https://github.com/maybe-eve/alpha-guide/tree/ja-master

Crowdin is now showing the translation ratio as 80%, but this is because I have to hack the translation to make the output file exported correctly.

**Comment by ayust** [comment]
Hm. It looks like something along the way caused page headers to not be handled properly - they're not being recognized by Hugo, for instance here:
http://alpha-guide-ja.netlify.com/rookie-fitting/

I'm going to investigate more to see if I can figure out why.

**Comment by ayust** [comment]
I think I figured out the issue - the `+++` at the start of the file must not have any whitespace trailing after it on the line. The files on your branch have a space or two after the `+++` before the new line, so it's not being recognized. I will fix it on my end.

**Comment by ayust** [comment]
Okay, I have implemented the fixes on my end, and this is now live at https://ja.thealphasguide.com - you can see the changes I had to make here:
https://github.com/ayust/alpha-guide/commits/translated-ja

**Comment by maybe-eve** [comment]
Thanks! The problem was that Crowdin ignores all the newlines in the translation. So I programmatically inserted <br />s, but it seems  Crowdin inserted white spaces after <br />s in turn.
I'll find the better way to treat with this issue. Again thank you very much!

**Comment by maybe-eve** [comment]
Hi,
I've just received the response from the Crowdin developers.

> Hello Mayvie,
> 
> A quick update regarding newlines in the header of your .md files.
> 
> We assume you use a static website engine HUGO to generate your sources and TOML is a front matter? As our developers say, you may switch to JSON instead of TOML, newlines structure is not important here. Please see more information on the page below:
> http://gohugo.io/content/front-matter/
> 
> Our developers have also created a separate task on the matter, since we plan to support custom .md files generated by different engines.
> 
> Thanks

If you could switch to JSON, it will be great.

**Comment by ayust** [comment]
Done in bf24948b1554cc719e1a4fed16cfbdc52e0b0da7 - hopefully that will help make your life a little easier (you may still need to do some tweaking though, such as for the bold tags in the icon fields).

**Comment by ayust** [comment]
For reference, here's the script I used to do the conversion. It doesn't try to be perfect because I didn't actually use a TOML parsing library, but since the front matter doesn't use any of the fancier features of TOML it works fine:

```python
#!/usr/bin/env python

import json
import os

JSON_CONFIG = {
  "indent": 2,
  "sort_keys": True,
  "ensure_ascii": False,
  "allow_nan": False,
  "separators": (",", ": "),
}

def strip_enclosing_quotes(val):
  if val[:1] == val[-1:] and val[:1] in ("'", '"'):
    val = val[1:-1]
  return val

def convert_toml_to_json(toml_input):
  # Note - this assumes that the input is valid TOML.
  # We don't try to handle invalid markup cases.
  fields = {}
  lines = toml_input.split("\n")
  for line in lines:
    key, equals_sign, value = line.partition("=")
    if not equals_sign:
      continue
    key = key.strip()
    value = value.strip()
    if not (key and value):
      continue
    key = strip_enclosing_quotes(key)
    value = strip_enclosing_quotes(value)
    fields[key] = value
  return json.dumps(fields, **JSON_CONFIG)

def fix_file(path):
  with open(path) as f:
    before = f.read()
  if not before.startswith("+++\n"):
    return False
  front_matter, ok, content = before[4:].partition("\n+++\n")
  if not ok:
    return False
  json_front_matter = convert_toml_to_json(front_matter)
  if json_front_matter is None:
    return False
  with open(path, "w") as f:
    f.write(json_front_matter)
    f.write("\n")
    f.write(content)
  return True


def main():
  fixed_count = 0
  for dirpath, _, filenames in os.walk(os.getcwd()):
    markdown_files = (f for f in filenames if f.endswith(".md"))
    markdown_paths = (os.path.join(dirpath, f) for f in markdown_files)
    fixed_count += sum(1 for p in markdown_paths if fix_file(p))
  print("{} files fixed.".format(fixed_count))

if __name__ == "__main__":
  main()
```

**Comment by ayust** [comment]
I have also converted the existing `translated-ja` branch and merged in master appropriately (see bfe7995). You may wish to pull from upstream.

**Comment by maybe-eve** [comment]
thanks and thanks!
